### ISSUE TITLE
Add support to docs preview for `examples` and `example` at the root of a schema

### ISSUE THREAD

**Issue by marbemac** [stoplightio/studio#381]
**User story.**

As a stoplight user, I can do use the `examples` and `example` properties at the root of a schema, and they will be rendered in docs.

**Is your feature request related to a problem?**

There are three ways (🙄) to add root level examples to a schema (embedded in openapi, or standalone file).

- `x-examples`, proprietary, used in openapi v2 when multiple examples are needed
- `example`
- `examples` 

**Describe the solution you'd like**

In our docs components (such as preview), we should support all three to power the example tabs. Note that I think we only need to choose one at a time, in this priority order: `examples`, `x-examples`, `example`.

**Additional context**

The form editor is out of scope here, we're just talking read only docs / preview here.

While we're at it, can we improve the example display so that it looks nice in light and dark mode? For example in screenshot below let's go with a transparent background (rather than dark) to match the white, and maybe reduce the padding?

<img width="732" alt="Screen Shot 2020-04-07 at 10 56 16 AM" src="https://user-images.githubusercontent.com/847542/78695959-6dd7be00-78c4-11ea-9298-25ad61b71c59.png">

**Comment by mallachari** [comment]
@marbemac just to clarify what is expected here. 
We already support `examples`+`example` in oas3 - they are combined together (default tab for `example` and named tabs for `examples`). Though it is not at the root of schema, rather adjacent to it. Should we support `x-examples` for oas3 as well? And also inside schema?
As for oas2 we support `examples` for responses (spec defines it only here), and `x-examples` for body (though it looks broken as it duplicates param). Again only adjacent to schema. Should we support all 3 here? Spectral shows error when using `example` for oas2 response (doesn't complain about body though).

**Comment by marbemac** [comment]
Hmm, `example` does not seem to work for me?

<img width="1158" alt="Screen Shot 2020-05-07 at 9 10 58 PM" src="https://user-images.githubusercontent.com/847542/81362771-40d72080-90a7-11ea-93e9-e772dd03a51c.png">

In any case, unless it is difficult I think we should support `example`, `examples`, and `x-examples` in schemas, wherever they are defined (oas2, oas3). The form editor won't necessarily produce all of that obviously - just talking elements / preview here.

**Comment by mallachari** [comment]
@philsturgeon I'm going to add support for schema examples to the docs. There are couple of ways we could do that. Here's how I think it could work. Could you see and share your thoughts?

### Schema models

We are currently picking one of two: `x-examples` and `examples`. 
`example` is valid keyword so we could pick one of three `examples` > `x-examples` > `example`.

### OAS3

OAS3 request and response objects allows `examples` and `example`. They also allow `examples` and `example` in schema. Properties defined in root object take precedence over those in schema.
Currently both request body and response build examples the same way which is concatenating `examples` and `example` from root object, ignoring schema.

Including schema we could have concatenation of
- `examples` from root or `examples` from schema if none in root
- `example` from root or `example` from schema if none in root
- `x-examples` from schema

### OAS2
**Request**
OAS2 doesn't define examples for requests. We use currently `x-examples` in root object.
Schema object may have `example` property.
Including schema we could have:
- `x-examples` from root object if exists or 
- one of two `x-examples` or `example` from schema, where `x-examples` > `example`

**Response**
OAS2 accepts `examples` for response.
Including schema we could have:
- `examples` from root object if exists or 
- one of two `x-examples` or `example` from schema, where `x-examples` > `example`

We could also concatenate all those values but it doesn't seem correct

**Comment by philsturgeon** [comment]
This is the worlds most confusing topic, so good job being extra clear. 

@marbemac this doesn't look like a valid piece of OpenAPI. 

![image](https://user-images.githubusercontent.com/67381/81608476-3fbb3180-93ce-11ea-9bef-0a6f4299f33f.png)

The `example-21` key looks like a named example, which is only possible in a Media Type Object and not a Schema Object.

@mallachari @marbemac Settle in for a long walk... 😞

There are [Media Type Objects](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#mediaTypeObject) which make up request and response content.

Inside there may be Schema Objects under the schema keyword. 

OAS3 Media Type Object `examples` and `example`
OAS3 Schema Object `examples` and `example`

_The media type object `examples` is keyed by an arbitrary string and then contains an object, the actual example sites inside the `value`, whilst the media type object `example` is a naked value._

OAS2 Media Type Object `examples` 
OAS2 Schema Object `x-example`

_The OAS2 media type object is keyed by a mime type (`application/json`) and does not have `value` to worry about.

@mallachari a few things I spotted in your reply:

1. OAS3 does not need any support for `x-examples`, as `examples` was added for Media Type object 
2. OAS3 does not support `examples` in the Schema Object, still. It's a JSON Schema keyword but not an OAS3 keyword.
3. i'm not sure when there would be a scenario where  `examples` > `x-examples` > `example`.

This is something the prism team have had to deal with. Think of two different types of examples completely separately.

1. Media Type Object Examples: If there are multiple, grab the first. If there is one, great stop.
2. If there is no media type object examples, we need to "generate" an example, by looking at the various properties in the Schema Object and taking all the examples, default values, first enum, etc.


ANYhow, this issue could probably be closed bc not supporting this OAS is correct behavior.

**Comment by mallachari** [comment]
> @marbemac this doesn't look like a valid piece of OpenAPI.
> 
> ![image](https://user-images.githubusercontent.com/67381/81608476-3fbb3180-93ce-11ea-9bef-0a6f4299f33f.png)
> 
> The `example-21` key looks like a named example, which is only possible in a Media Type Object and not a Schema Object.

Marc was referring to standalone schema file. `example-21` is rather a value of example property, not its name.

> 2. OAS3 does not support `examples` in the Schema Object, still. It's a JSON Schema keyword but not an OAS3 keyword.

Wait but earlier you wrote 
> OAS3 Schema Object examples and example

then I'm assuming that we can still use `examples` from schema object if there are no media type examples?

> 3. i'm not sure when there would be a scenario where  `examples` > `x-examples` > `example`.
> 

That would refer to standalone schema object file. Currently we are supporting `examples` and `x-examples` there, picking one of them.

Thanks for looking into this. I think I get how it could look like now.

**Comment by philsturgeon** [comment]
@mallachari sorry good catch, I've updated my comment in regards to OAS3 Schema Object `examples`, specifically:

>  OAS3 Schema Object `examples` (sometimes supported bc JSON Schema has it but usually not as its not in OAS3) and `example`.

Next bit:

> Marc was referring to standalone schema file. example-21 is rather a value of example property, not its name.

The example provided does not match up to the properties mentioned, so maybe im just a little confused about what is being requested.

```
definitions: 
  # A: not sure what this is meant to look like
  invalid-example-object:
    type: object
    properties:
      exclude:
        type: array
        items: 
          type: string
      formats:
        type: string
    example: 
      example-21: {} # wait what is example-21

  # B: this would be a valid schema object example for OAS3 
  oas3-valid-example-object:
    type: object
    properties:
      exclude:
        type: array
        items: 
          type: string
      formats:
        type: string
    example: # the example object should be a valid object of the schema its describing 🖕🏻
      exclude: [foo, bar]
      formats: [something]

  # C: An alternative way to do exactly the same thing, 
  # placing example values in their properties instead of an object
  oas3-valid-example-properties:
    type: object
    properties:
      exclude:
        type: array
        items: 
          type: string
        example: [foo, bar]
      formats:
        type: string
        example: [something]
  
  # D: Let's do these last two examples in OAS2
  oas2-valid-example-object:
    type: object
    properties:
      exclude:
        type: array
        items: 
          type: string
      formats:
        type: string
    x-example: # the example object should be a valid object of the schema its describing 🖕🏻
      exclude: [foo, bar]
      formats: [something]

  # E: An alternative way to do exactly the same thing, 
  # placing example values in their properties instead of an object
  oas2-valid-example-properties:
    type: object
    properties:
      exclude:
        type: array
        items: 
          type: string
        x-example: [foo, bar]
      formats:
        type: string
        x-example: [something]
```

Nearly on the same page!

A looks like this in Prism, which is correct: 

<img width="1040" alt="Screen Shot 2020-05-12 at 18 02 55" src="https://user-images.githubusercontent.com/67381/81723650-13a9ba00-947b-11ea-87be-1e8146c6555f.png">

B looks like this in Prism, which is correct:

<img width="1042" alt="image" src="https://user-images.githubusercontent.com/67381/81724595-8b2c1900-947c-11ea-9b40-53ed7afef85b.png">

C looks the same, which is expected: 

<img width="1093" alt="image" src="https://user-images.githubusercontent.com/67381/81724785-d9d9b300-947c-11ea-8745-7228e5f6e56a.png">

So, example A might be a tad confusing because it _looks_ like named examples. If we use B or C as examples for this feature, we might be a little more sure we're on the same page.

If you can make the examples look like whatever Prism is doing, you shouldn't go far wrong. We put a _lot_ of effort into those examples.

Check out the [Prism: HTTP Mocking Guide](https://stoplight.io/p/docs/gh/stoplightio/prism/docs/guides/01-mocking.md).

It'll either use a request/response example, or failing that it'll generate one from whatever combination of Schema Object examples are found. 

JSON Schema Faker is good at this.

**Comment by rossmcdonald** [comment]
What release is this change going into? Is it possible to get it into v1, or is this just for v2?

**Comment by mallachari** [comment]
@rossmcdonald it is going into v2 but I plan to add it to v1 as well

**Comment by rossmcdonald** [comment]
@mallachari So is this fixed in v2 already? I may not be understanding. If so we can probably close.

**Comment by mallachari** [comment]
@rossmcdonald yes, done in v2. I'm going to add it to v1 too, though it needs different approach as it's entirely in studio IIRC.

**Comment by mallachari** [comment]
Closing as it's done for v2. 
Final approach taken:

Schema examples support for OAS2/OAS3/standalone schema files according to what is allowed within spec.

**OAS3**
OAS3 supports `examples` and `example` for media type object and `example` for schema object. We use `example` from schema object if media type `example` is not provided. Extensions are ignored.

**OAS2**
OAS2 supports `examples` for media type response object and nothing for request - hence we use `x-examples` extension (as we did before).
Schema object supports only `example` - we extend it and allow also `x-examples`.
Schema object examples are used if media type examples are not provided.

**JSON Schema** (standalone schema files)
JSON Schema supports `examples` only in v6 and v7. We don't have any way to tell which version is used so we allow `x-examples` extension along with `x-example`. 
Examples are combined together.
